web/src/tenor.ts

```diff
@@ -65,6 +67,64 @@ function handle_gif_click(img_element: HTMLElement): void {
     hide_tenor_popover();
 }
 
+function focus_gif_at_index(index: number): void {
+    if (index > last_gif_index) {
+        // Just trigger focus on the compose box, because there are no GIFs below.
+        $("#compose-textarea").trigger("focus");
+        return;
+    }
+    if (index < 0) {
+        assert(tenor_popover_instance !== undefined);
+        const $popper = $(tenor_popover_instance.popper);
+        // Just trigger focus on the search input because there are no GIFs above.
+        $popper.find("#gif-search-query").trigger("focus");
+        return;
+    }
+
+    const $target_gif = $(`img.tenor-gif[data-gif-index='${index}']`);
+    $target_gif.trigger("focus");
+}
+
+function handle_keyboard_navigation_on_gif(e: JQuery.KeyDownEvent): void {
+    assert(e.currentTarget instanceof HTMLElement);
+    const key = e.key;
+    const is_alpha_numeric = /^[a-zA-Z0-9]$/i.test(key);
+    if (is_alpha_numeric) {
+        // This implies that the user is focused on some GIF
+        // but wants to continue searching.
+        assert(tenor_popover_instance !== undefined);
+        const $popper = $(tenor_popover_instance.popper);
+        $popper.find("#gif-search-query").trigger("focus");
+        return;
+    }
+    if (key === "Enter" || key === " " || key === "Spacebar") {
+        // Meant to avoid page scroll on pressing space.
+        e.preventDefault();
+        handle_gif_click(e.currentTarget);
+        return;
+    }
+
+    const curr_gif_index = Number.parseInt(e.currentTarget.dataset["gifIndex"]!, 10);
+    switch (key) {
+        case "ArrowRight": {
+            focus_gif_at_index(curr_gif_index + 1);
+            break;
+        }
+        case "ArrowLeft": {
+            focus_gif_at_index(curr_gif_index - 1);
+            break;
+        }
+        case "ArrowUp": {
+            focus_gif_at_index(curr_gif_index - 3);
+            break;
+        }
+        case "ArrowDown": {
+            focus_gif_at_index(curr_gif_index + 3);
```

Comment:
I think rather than hardcoding 3, you're going to want to read the numbers of columns the picker has in the DOM. (I imagine with a narrow window, we won't fit all 3?)

Reply:
Yeah that's fine, basically when we do the deferred "Make it bigger, which requires variable width" feature.